Issue #8 feat: Refactor list view components by adding prompts to the differen…

RossaMania opened the pull request:
…t forms. Conditionally render buttons to go to and from the "/list" and the "/manage-list" views. Add navigation functionality with useNavigate hook.

Co-authored-by: Falak  <zahrafalak@users.noreply.github.com>

_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._

## Description

<!-- What does this code change? Why did I choose this approach? Did I learn anything worth sharing? Reminder: This will be a publicly facing representation of your work (READ: help you land that sweet dev gig). -->

Refactor list view components by adding prompts to the different forms. Conditionally render buttons to go to and from the "/list" and the "/manage-list" views. Add navigation functionality with useNavigate hook.

When a user starts a new list, they need a little guidance on how to get started. We added instructions in the "/list" view when the list is empty. We also added a button that takes the user over to the "/manage-list" view. Additional prompts were added as well. These included instructions on adding an item and picking when the user would want to buy that item next. Also, a prompt and button to go back to the "/list" view was added as well.

We wanted to use welcoming and conversational language, but also provide firm instructions on what the user can and should do. Like a "digital assistant" but actually helpful! 😸 

We already have React Router installed, so we kept it simple by using the useNavigate hook and click handlers to guide the user from page to page!

## Related Issue

<!-- If you write "closes" followed by the Github issue number, it will automatically close the issue for you when the PR merges -->
closes #8 

## Acceptance Criteria

<!-- Include AC from the Github issue -->
 

- [ ] The list view, when there are no items to display, should show a prompt (e.g., a button) for the user to add their first item

## Type of Changes

Use one or more labels to help your team understand the nature of the change(s) you’re proposing. E.g., `bug fix` or `enhancement` are common ones.

`enhancement` `user experience` `helpful feature`

## Updates

### Before

<!-- If UI feature, take provide screenshots -->
![image](https://github.com/user-attachments/assets/963924e2-f64c-4fcf-8557-5b6b3114f0f9)
![image](https://github.com/user-attachments/assets/dbd3d524-531d-4dd3-9f62-51b71a45c326)


### After

<!-- If UI feature, take provide screenshots -->
![image](https://github.com/user-attachments/assets/c76f0d5a-c389-4a75-8a2d-706a59d3e7ed)
![image](https://github.com/user-attachments/assets/9d009bd4-6b7b-4fa1-ae40-0cb9ad0ae753)
![image](https://github.com/user-attachments/assets/fe3129af-4c4d-4163-ba3b-079b0896a618)


## Testing Steps / QA Criteria

<!-- Provide steps the other cohort members and mentors need to follow to properly test your additions. -->
Click this preview link for this PR! 
https://tcl-77-smart-shopping-list--pr32-fz-rc-feat-add-promp-3zw3l7nm.web.app/
Remember, our app uses Google auth. Please sign in with your Google account!
1. Make sure you see the message "Hello from the home(/) page!"
2. Type in the name of your new list in the "Name Your List" text field.
3. Click that "Create List" button.
4. This should take you to the "/list" view. If you don't see the "Hello from the /list page!" please click on that big blue List link at the bottom of the app screen.
5. Read the prompt, and look for the "Get started!" button. Click the "Get started!" button.
6. You should see the "Hello from the /manage-list page!" and the new, helpful prompts we have written!
7. Follow the prompts, click the buttons, and be sure to try out the "Let's go!" navigation button on the /manage-list page, as well as the "Add items" button on the /list page (only visible once items are added) to try out that smooth navigation between pages! 😄

bbland1 commented on src/views/authenticated/List.tsx:
```diff
@@ -16,17 +19,47 @@ export function List({ data: unfilteredListItems }: Props) {
 				Hello from the <code>/list</code> page!
 			</p>
 
+			{unfilteredListItems.length === 0 && (
+				<div>
+					<h2>Your list is ready!</h2>
+					<h3>
+						It looks like you haven’t added any items yet.
+						<br />
+						Let’s get started by adding your first item.
+					</h3>
+					<button
+						onClick={handleAddItemsClick}
+						aria-label="Navigate to add items to your list"
+					>
+						{"Get started!"}
+					</button>
+				</div>
+			)}
```
If the button logic is centralized and the 2 states for when there is no data and is data are still wanted I think this could get changed to a terneary instead of the short eval and this could handled the change in the prompt. It would allow the logic for this to be in one place instead of across the 2 component areas.

```suggestion
			{unfilteredListItems.length === 0 ? (
				<div>
					<h2>Your list is ready!</h2>
					<h3>
						It looks like you haven’t added any items yet.
						<br />
						Let’s get started by adding your first item.
					</h3>
					<button
						onClick={handleAddItemsClick}
						aria-label="Navigate to add items to your list"
					>
						{"Get started!"}
					</button>
				</div>
			) : (
                <h3>Want to add more items to your list?</h3>
            )}
```

eternalmaha replied:
Great logical thought @bbland1 !!! I like that you guys incorporated semantic HTML with the different headers. I wanted to add on to that by suggesting we replace the <div> elements with a <section> for improved semantic HTML

RossaMania replied:
Nice!

I was thinking the other night ,"I've been using divs a lot in this project", but it was late at night and I couldn't remember any other elements to use besides main or body! 🤣 🤣 🤣 🤣 🤣

bbland1 commented on src/views/authenticated/List.tsx:
```diff
@@ -16,17 +19,47 @@ export function List({ data: unfilteredListItems }: Props) {
+					<h3>
+						It looks like you haven’t added any items yet.
+						<br />
+						Let’s get started by adding your first item.
```
thought: It may be itersting to think if we could make a shorter promot (so maybe more of a nitpick I'm sorry! 😅) but it maybe like:

"No items? Let's add the first one!"

...Well that felt cheesy, I am sorry for how unsmooth that was lol 😂

RossaMania replied:
I can see where you are coming from! We meant for the prompts to be conversational in tone and helpfully encouraging.
I actually agree though, that this prompt could be shorter:

You haven't added any items yet.
br
Let's get started by adding your first item!

zahrafalak replied:
I align with this refactor as well! :point_up:

bbland1 commented on src/views/authenticated/List.tsx:
```diff
@@ -16,17 +19,47 @@ export function List({ data: unfilteredListItems }: Props) {
 				Hello from the <code>/list</code> page!
 			</p>
 
+			{unfilteredListItems.length === 0 && (
```
praise: I love the short eval instead of a bunch of if-elses. it's easiery for me to follow personally.

bbland1 commented on src/views/authenticated/ManageList.tsx:
```diff
@@ -154,6 +163,8 @@ export function ManageList({ listPath }: Props) {
 						</button>
 					</form>
+					<h4>Let&apos;s look at your list!</h4>
+					<button onClick={navigateToListPage}>{"Let's go!"}</button>
```
praise: I like adding a button to get back to the list!

eternalmaha replied:
The button to get back to the list is definitely a great added feature that improves user experience and functionality. I was thinking, from our  group conversation last night about accessibility, it would be a great idea to rename the button perhaps to "View List" instead of "Let's go!"! so it speaks more to the buttons purpose :) and then perhaps the text above it "Let's look at your list" would be optional? I do like how it's so conversational though!

RossaMania commented:
I should have some free time tomorrow.

Does anybody have a problem if I try to refactor this ManageList.tsx view and make a AddItemForm.tsx component to render inside that ManageList view?

eternalmaha commented:
@RossaMania I think refactoring the ManageList view to render an addItem component is definitely needed and helpful at some point! I remember that was a conversation we were having in the previous issue and as we both were building the add item feature, since the application is getting so large, and ManageList has a few different functions now

RossaMania commented on src/views/authenticated/List.tsx:
```diff
@@ -16,17 +19,47 @@ export function List({ data: unfilteredListItems }: Props) {
+					</button>
+				</div>
+			)}
+
```
So basically, making some sort of re-usable "Button" component with a name prop (and probably a link prop)?

Feels responsible.

bbland1 replied:
That's definitely an option to make a reusable component 🤔. I was more thinking of taking the two instances of the `button` in the 2 stops and combining them and have just one button in the page that has a dynamic name really. So not really a new component just minimizing the duplication.

tannaurus replied:
One thing I really try to resist is "DRY" (don't repeat yourself) for the sake of "DRY"

I'm a huge fan of repeating yourself. It always for flexibility when you need it. Only when there are very clear signs that you're actually repeating the same thing 2-3 times, or its clear that in the very near future that will be the case, will I tend to refactor things out to be there own component. In the case of this button, I'd argue the complexity of what we're trying to do here (display the right button) has a meaningful increase now that it's try to handle its own logic internally. The code becomes much simpler with two big branches: one view that shows no items, and an alternative view that shows with some items.

RossaMania commented:
> @RossaMania I think refactoring the ManageList view to render an addItem component is definitely needed and helpful at some point! I remember that was a conversation we were having in the previous issue and as we both were building the add item feature, since the application is getting so large, and ManageList has a few different functions now

![image](https://github.com/user-attachments/assets/d3ec226f-89d6-45f9-b74f-2d3af66ebd03)

There's probably room for like a PurchaseTimelineForm or BuyAgainForm or something, but that may be too granular--too specific--since it's related to adding an item to a list. Either way, a fantastic "tomorrow problem"!

tannaurus commented on src/views/authenticated/List.tsx:
```diff
@@ -16,17 +19,47 @@ export function List({ data: unfilteredListItems }: Props) {
+						Let’s get started by adding your first item.
+					</h3>
+					<button
+						onClick={handleAddItemsClick}
```
A saying I _love_ is "avoid shallow abstractions" - a phrase I personally grabbed from John Ousterhout who authored the book "A Philosophy of Software Design" - this talk is a great introduction to that: https://www.youtube.com/watch?v=bmSAYlu0NcY

Try reading this from the perspective of someone who is first reading this component for the first time, or from the perspective of yourself reading this 6 months from now: can you figure out what `handleAddItemsClick` means? You'd probably need to click into the function, only to find out it's actually a wrapper around a single function call. It's actually a lot easier to communicate what this button click does by just inlining the function directly. If that function gets sufficiently complex, moving it into its own function starts to make a lot more sense

tannaurus commented on src/views/authenticated/List.tsx:
```diff
@@ -16,17 +19,47 @@ export function List({ data: unfilteredListItems }: Props) {
 	return (
 		<>
 			<p>
 				Hello from the <code>/list</code> page!
```
Do we need this anymore?

I ask because I think this component would become a lot easier to read if we return early instead of inlining conditions inside the tsx.

Something like this:

```
if (unfilteredListItems.length === 0) {
  return ( 
     <> ... "your list is ready" </>
  )
}

return (
  <>... filter list, etc <>
)
```

tannaurus reviewed (approved):
Great work ya'll! Left a couple thoughts that you can feel free to do with what you wish 🙂

RossaMania commented:
Falak and I met and were able to discuss the feedback!
- We used in-line functions for navigation, instead of click handlers!
- We shortened up some prompts for readability!
-  We changed some button names for clarity as well--for both users and accessibility!
-  We changed some div elements to sections HTML elements!

alex-andria reviewed (approved):
Nice job Ross and Falak! I love all the different user friendly prompts between empty list pages and prompts for adding your next list. Clean use of your form functionality as well. 🧠

One thing to maybe think about in the future is if you wanted the instructional prompt "Next, pick when you plan on buying this item again!" and "Let's go look at your list!" to always be present - which is how it is currently even after multiple items are added. Or if you'd like it to only be present until after the first item is added. 

Otherwise, great job to you both for addressing all the other comments so quickly!

![image](https://github.com/user-attachments/assets/21991637-84ec-4ddd-ba8c-63c05a032996)
